libs/api/ebpf_api.cpp

```diff
@@ -1510,6 +1517,115 @@ ebpf_object_open(
     return result;
 }
 
+static inline bool
+_ebpf_is_map_in_map(ebpf_map_t* map)
+{
+    if (map->map_definition.type == BPF_MAP_TYPE_HASH_OF_MAPS ||
+        map->map_definition.type == BPF_MAP_TYPE_ARRAY_OF_MAPS) {
+        return true;
+    }
+
+    return false;
+}
+
+static ebpf_result_t
+_ebpf_object_reuse_map(_In_ ebpf_map_t* map)
+{
+    ebpf_result_t result = EBPF_SUCCESS;
+    fd_t inner_map_info_fd = ebpf_fd_invalid;
+
+    // Check if a map is already present with this pin path.
+    fd_t map_fd = ebpf_object_get(map->pin_path);
+    if (map_fd == ebpf_fd_invalid) {
+        return EBPF_SUCCESS;
+    }
+
+    // Validate that the existing map definition matches with this new map.
+    struct bpf_map_info info;
+    uint32_t info_size = (uint32_t)sizeof(info);
+    result = ebpf_object_get_info_by_fd(map_fd, &info, &info_size);
+    if (result != EBPF_SUCCESS) {
+        result = EBPF_INVALID_ARGUMENT;
+        goto Exit;
+    }
+
+    if (info.type != map->map_definition.type || info.key_size != map->map_definition.key_size ||
+        info.value_size != map->map_definition.value_size || info.max_entries != map->map_definition.max_entries) {
+        result = EBPF_INVALID_ARGUMENT;
+        goto Exit;
+    }
+
+    // Extra checks for map-in-map.
+    if (_ebpf_is_map_in_map(map)) {
+        ebpf_map_t* inner_map = map->inner_map;
+        ebpf_assert(inner_map != nullptr);
+        ebpf_id_t inner_map_id = info.inner_map_id;
+
+        if (inner_map_id == EBPF_ID_NONE) {
+            // The original map is pinned but its template is not initialized yet.
+            // This is not supported for now.
+
+            // Workaround: Check if there are any entries in this map. If present,
+            // use the map entry to get the inner map template.
+            uint32_t key;
+            result = ebpf_map_get_next_key(map_fd, nullptr, &key);
+            if (result != EBPF_SUCCESS) {
+                result = EBPF_INVALID_ARGUMENT;
+                goto Exit;
+            }
+
+            // Got a key. Query the value. That will be the map id of the inner map.
+            result = ebpf_map_lookup_element(map_fd, &key, &inner_map_id);
+            if (result != EBPF_SUCCESS || inner_map_id == EBPF_ID_NONE) {
+                result = EBPF_INVALID_ARGUMENT;
+                goto Exit;
+            }
+        }
+
+        if (inner_map->map_definition.pinning == PIN_GLOBAL_NS) {
+            // Inner map cannot be auto-pinned.
```

Comment:
Why not?

Reply:
The inner map here is just a template, and not an actual map, so I think there is no need to pin it. App can pin the actual inner map after it has created the inner map.

Reply:
"cannot be" sounds like it's impossible, which is different from "not necessary".  If you mean "There is no need to pin it", then say that.

Reply:
Removed this restriction for inner maps, and updated test cases to test pinning of inner maps.